### ISSUE TITLE
Unnecessary Pyright: Method declaration "kdm_file_path" is obscured by a declaration of the same name [reportGeneralTypeIssues]

### ISSUE THREAD

**Issue by John15321** [microsoft/pyright#2284]
**Describe the bug**
Im writing a class (I posted its whole code below). I added a setter getter and its complaining that:

```
Method declaration "kdm_file_path" is obscured by a declaration of the same name
```

Which doesn't seem to be true (If my code is wrong correct me).
Im using [Lunar Vim](https://www.lunarvim.org/#opinionated), if I open the same code in Visual Studio Code this error does not exist there
Lunar Vim devs told me its an issue with` pyright` which is used by `LunarVim`

Vim:
![image](https://user-images.githubusercontent.com/56229312/132691561-914b8753-1e1b-4a9f-94a1-2c7a3fa4a080.png)

VScode:
![image](https://user-images.githubusercontent.com/56229312/132691625-f281e454-1e18-4325-9aa5-197dfbc9ff30.png)

**To Reproduce**
Just open a python file with the code below

**Expected behavior**
It shouldnt give that error

**Screenshots or Code**
```python
class KDM:
    def __init__(self, kdm_file_path: Path) -> None:
        self._kdm_file_path: Path = Path()

        self.kdm_file_path: Path = kdm_file_path

    @property
    def kdm_uuid(self) -> str:
        return self.kdm_file_name.replace(KDM_EXTENSION, "")

    @property
    def kdm_file_name(self) -> str:
        return self._kdm_file_path.name

    @property
    def kdm_file_path(self) -> Path:
        return self._kdm_file_path

    @kdm_file_path.setter
    def kdm_file_path(self, kdm_file_path: Path):
        if not kdm_file_path.exists():
            raise Exception(f"The kdm file path: {kdm_file_path}. Does not exist.")
        self._kdm_file_path = kdm_file_path
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

**Additional context**

**Comment by erictraut** [comment]
The problem is that the symbol `kdm_file_path` has been declared in two different incompatible ways. It is being declared as a property, but it's also being annotated as a variable with type `Path` (in the `__init__` method). If you change `self.kdm_file_path: Path = kdm_file_path` to `self.kdm_file_path = kdm_file_path`, the error will go away.

**Comment by John15321** [comment]
> The problem is that the symbol `kdm_file_path` has been declared in two different incompatible ways. It is being declared as a property, but it's also being annotated as a variable with type `Path` (in the `__init__` method). If you change `self.kdm_file_path: Path = kdm_file_path` to `self.kdm_file_path = kdm_file_path`, the error will go away.

Ohh okay. That makes sense, thanks! Though doesn't that mean that VScode isn't showing all the errors/warnings that it should?

**Comment by erictraut** [comment]
> doesn't that mean that VScode isn't showing all the errors/warnings that it should?

I'm not sure what you mean. You received an error indicating that a declaration was obscuring a declaration of the same name, right? The error is reported on the second (incompatible) declaration encountered within the class definition. Are you saying that you would like to see the error reported for the first of the two incompatible declarations in this instance? Or you want to see errors in both places? Or something else?

**Comment by John15321** [comment]
> > doesn't that mean that VScode isn't showing all the errors/warnings that it should?
> 
> I'm not sure what you mean. You received an error indicating that a declaration was obscuring a declaration of the same name, right? The error is reported on the second (incompatible) declaration encountered within the class definition. Are you saying that you would like to see the error reported for the first of the two incompatible declarations in this instance? Or you want to see errors in both places? Or something else?

What I mean is that if the error message that I got in vim is justified, shouldn't it also appear in VScode?

**Comment by bschnurr** [comment]
I think issue is the error `"Method declaration "kdm_file_path" is obscured by a declaration of the same name"` when we set typechecking mode to basic.

`"python.analysis.typeCheckingMode": "basic",`

**Comment by erictraut** [comment]
In pylance (which is the extension you're using in VS Code), type checking diagnostics are disabled by default. That's because most pylance users are more interested in completion suggestions and other language features and less interested in type checking errors. As @bschnurr mentioned, you can enable type checking in pylance, and that should give you the same results that you see when you run pyright in vim.

**Comment by John15321** [comment]
> In pylance (which is the extension you're using in VS Code), type checking diagnostics are disabled by default. That's because most pylance users are more interested in completion suggestions and other language features and less interested in type checking errors. As @bschnurr mentioned, you can enable type checking in pylance, and that should give you the same results that you see when you run pyright in vim.

Got it! Thank you very much! :D
